src/manifolds/MultinomialDoublyStochastic.jl

```diff
@@ -0,0 +1,286 @@
+            r,
+            "The matrix $(X) is not a tangent vector to $(p) on $(M), since its rows do not sum up to zero.",
+        )
+    end
+    return nothing
+end
+
+function decorated_manifold(::MultinomialDoubleStochasticMatrices{N}) where {N}
+    return Euclidean(N, N; field = ℝ)
+end
+
+embed!(::MultinomialDoubleStochasticMatrices, q, p) = copyto!(q, p)
+embed!(::MultinomialDoubleStochasticMatrices, Y, p, X) = copyto!(Y, X)
+
+@doc raw"""
+    inner(M::MultinomialDoubleStochasticMatrices{n}, p, X, Y) where {n}
+
+Compute the inner product on the tangent space at $p$, which is the elementwise
+inner product similar to the [`Hyperbolic`](@ref) space, i.e.
+
+````math
+    \langle X, Y \rangle_p = \sum_{i,j=1}^n \frac{X_{ij}Y_{ij}}{p_{ij}}.
+````
+"""
+function inner(::MultinomialDoubleStochasticMatrices, p, X, Y)
+    # to avoid memory allocations, we sum in a single number
+    d = zero(Base.promote_eltype(p, X, Y))
+    @inbounds for i in eachindex(p, X, Y)
+        d += X[i] * Y[i] / p[i]
+    end
+    return d
+end
+
+@doc raw"""
+    manifold_dimension(M::MultinomialDoubleStochasticMatrices{n}) where {n}
+
+returns the dimension of the [`MultinomialDoubleStochasticMatrices`](@ref) manifold
+namely
+````math
+\operatorname{dim}_{\mathcal{DP}(n)} = (n-1)^2.
+````
+"""
+@generated function manifold_dimension(::MultinomialDoubleStochasticMatrices{n}) where {n}
+    return (n - 1)^2
+end
+
+@doc raw"""
+    project(M::MultinomialDoubleStochasticMatrices{n}, p, Y) where {n}
+
+Project `Y` onto the tangent space at `p` on the [`MultinomialDoubleStochasticMatrices`](@ref) `M`, return the result in `X`.
+The formula reads
+````math
+    \operatorname{proj}_p(Y) = Y - (α\mathbf{1}_n^{\mathrm{T}} + \mathbf{1}_nβ^{\mathrm{T}}) ⊙ p,
+````
+where $⊙$ denotes the Hadamard or elementwise product and $\mathbb{1}_n$ is the vector of length $n$ containing ones.
+The two vectors $α,β ∈ ℝ^{n×n}$ are computed as a solution (typically using the left pseudo inverse) of
+````math
+    \begin{pmatrix} I_n & p\\p^{\mathrm{T}} & I_n \end{pmatrix}
+    \begin{pmatrix} α\\ β\end{pmatrix}
+    =
+    \begin{pmatrix} Y\mathbf{1}\\Y^{\mathrm{T}}\mathbf{1}\end{pmatrix},
+````
+where $I_n$ is teh $n×n$ unit matrix and $\mathbf{1}_n$ is the vector of length $n$ containing ones.
+
+"""
+project(::MultinomialDoubleStochasticMatrices, ::Any, ::Any)
+
+function project!(::MultinomialDoubleStochasticMatrices{n}, X, p, Y) where {n}
+    ζ = [I p; p I] \ [sum(Y, dims = 2); sum(Y, dims = 1)'] # Formula (25) from 1802.02628
+    return X .= Y .- (repeat(ζ[1:n], 1, 3) .+ repeat(ζ[(n + 1):end]', 3, 1)) .* p
+end
+
+@doc raw"""
+    project(
+        M::MultinomialDoubleStochasticMatrices,
+        p;
+        maxiter = 100,
+        tolerance = eps(eltype(p))
+    )
+
+project a matrix `p` with positive entries applying Sinkhorn's algorithm.
+"""
+function project(M::MultinomialDoubleStochasticMatrices, p; kwargs...)
+    q = allocate_result(M, project, p)
```

Comment:
Do we actually need to implement this method?

Reply:
I wanted to pass down the `kwargs`, since usually we don't do that for project.

Reply:
Right, it makes sense. Could you add a comment here in the code with this explanation?

Reply:
Of course.